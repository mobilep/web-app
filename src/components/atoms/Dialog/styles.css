@import url('../../../../node_modules/dialog-polyfill/dialog-polyfill.css');

dialog {
	--small-border-radius: 5px;

	border: none;
	border-radius: 5px;
	display: flex;
	justify-content: center;
	align-items: center;
}

dialog::backdrop,
dialog + .backdrop {
	background-color: rgba(135, 135, 135, .7);
}

.Dialog-close-button {
	cursor: pointer;
	width: 31px;
	height: 31px;
	border: none;
	padding: 0;
	position: absolute;
	top: 10px;
	right: 10px;
	background: none;
	color: var(--grey3);
}

.Dialog-close-button:focus {
	border: none;
}

@supports (backdrop-filter: blur()) or (-webkit-backdrop-filter: blur()) {
	dialog::backdrop,
	dialog + .backdrop {
		background-color: rgba(135, 135, 135, .5);
		opacity: 1;
		backdrop-filter: blur(3px);
	}
}
