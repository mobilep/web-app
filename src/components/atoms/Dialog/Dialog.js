import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import constants from '../../../constants';

import { CloseIcon } from '../../../icons';

import './styles.css';

const { common: { landscapeMediaQuery } } = constants;

export class Dialog extends Component {

	constructor (props) {
		super(props);
		if (!this.isDialogSupported) this.loadPolyfill();
	}

	isDialogSupported = window.HTMLDialogElement;
	dialogRef = this.props.forwardedRef || React.createRef();

	state = {
		isModalReady: this.isDialogSupported || false,
	}

	componentDidMount () {
		if (this.isDialogSupported) this.setUpModal();
		this.mql = matchMedia(landscapeMediaQuery);
		this.mql.addListener(this.handleOrientationChange);
		this.handleOrientationChange(this.mql);
	}

	componentWillUnmount () {
		const { onClose } = this.props;
		const currentDialog = this.currentDialog();
		if (currentDialog) currentDialog.removeEventListener('close', onClose);
	}

	currentDialog () {
		return this.dialogRef.current;
	}

	handleOrientationChange = ({ matches }) => {
		if (matches) {
			this.setState({ landscapeMode: true });
		} else {
			this.setState({ landscapeMode: false });
		}
	}

	handleClick = () => {
		this.closeModal();
	};

	closeModal () {
		this.currentDialog().close();
	}

	setUpModal = () => {
		const { visible, onClose } = this.props;
		this.currentDialog().addEventListener('close', onClose);
		if (visible) {
			this.showModal();
		}
	}

	showModal () {
		this.currentDialog().showModal();
	}

	async loadPolyfill () {
		const dialogPolyfill = await import('dialog-polyfill');
		this.setState({ isModalReady: true }, () => {
			const currentDialog = this.currentDialog();
			dialogPolyfill.registerDialog(currentDialog);
			this.setUpModal();
		});
	}

	getDialogClassName ({ className }, { landscapeMode }) {
		return classNames({
			'Dialog-landscapeMode': landscapeMode,
		}, className);
	}

	renderModal ({ children }) {
		const { closeButton } = this.props;

		return (
			<dialog
				className={this.getDialogClassName(this.props, this.state)}
				ref={this.dialogRef}>
				{closeButton &&
					<button
						className='Dialog-close-button'
						onClick={this.handleClick}
					>
						<CloseIcon />
					</button>}
				{children}
			</dialog>
		);
	}

	render () {
		const { isModalReady } = this.state;
		if (!isModalReady) return null;

		return (
			!this.isDialogSupported
				? ReactDOM.createPortal(
					this.renderModal(this.props),
					document.body)
				: this.renderModal(this.props)
		);
	}
}

Dialog.defaultProps = {
	onClose: () => {},
	closeButton: true,
};

Dialog.propTypes = {
	children: PropTypes.any,
	className: PropTypes.string,
	closeButton: PropTypes.bool,
	forwardedRef: PropTypes.any,
	onClose: PropTypes.func,
	visible: PropTypes.bool,
};

/* eslint-disable-next-line react/no-multi-comp */
const _Dialog = (props, ref) => <Dialog {...props} forwardedRef={ref} />;
_Dialog.displayName = 'Dialog';

// NOTE: Resolves an issue where enzyme@^3.3.0 and enzyme-adapter-react-16@^1.1.1 don't support `forwardRef`.
export default process.env.NODE_ENV === 'test' ? Dialog : React.forwardRef(_Dialog);
