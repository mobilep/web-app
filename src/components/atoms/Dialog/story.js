import React, { Fragment } from 'react';
import { storiesOf } from '@storybook/react';
import { text } from '@storybook/addon-knobs/react';

import Dialog from './Dialog';
import Button from '../Button';

const dialogRef = React.createRef();

storiesOf('1. Atoms / Dialog', module)
	.add('with text', () =>
		(<Fragment>
			<Dialog ref={dialogRef}>
				<div style={{ marginTop: '20px' }}>{text('Text', 'Hello! This is dialog component! Enjoy!')}</div>
			</Dialog>
			<Button primary onClick={() => dialogRef.current.showModal()}>Open dialog</Button>
		</Fragment>
		)
	);
