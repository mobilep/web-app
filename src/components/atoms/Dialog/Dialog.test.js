import React from 'react';
import { mount } from 'enzyme';
import dialogPolyfill from 'dialog-polyfill';
import mockMatchMedia from '../../../helpers/test/matchMedia';

import Dialog from './Dialog';

jest.mock('dialog-polyfill');

describe('Atom - Dialog', () => {
	beforeAll(mockMatchMedia);

	it('Should render', () => {
		global.HTMLDialogElement = true;
		const wrapper = mount(
			<Dialog>
				Dialog content
			</Dialog>
		);
		expect(wrapper).toMatchSnapshot();
	});

	describe('When HTMLDialogElement is not supported', () => {

		it('Should use polyfill if ', async () => {
			global.HTMLDialogElement = false;
			await mount(
				<Dialog>
					Dialog content
				</Dialog>
			);
			// Forces below statement to wait for polyfill to load.
			await import('dialog-polyfill');
			expect(dialogPolyfill.registerDialog).toHaveBeenCalledTimes(1);
		});

	});

});
