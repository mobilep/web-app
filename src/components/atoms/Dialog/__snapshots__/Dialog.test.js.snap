// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom - Dialog Should render 1`] = `
<Dialog
  closeButton={true}
  onClose={[Function]}
>
  <dialog
    className="Dialog-landscapeMode"
  >
    <button
      className="Dialog-close-button"
      onClick={[Function]}
    >
      <CloseIcon>
        <SVGWrapper
          color="currentColor"
          height={31}
          width={31}
        >
          <svg
            className="svgWrapper"
            version="1.1"
            viewBox="0 0 31 31"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
          >
            <g
              fill="none"
              fillRule="evenodd"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            >
              <path
                d="M21.333 21.333l-6.666-6.666L21.333 8"
              />
              <path
                d="M8 21.333l6.667-6.666L8 8"
              />
            </g>
          </svg>
        </SVGWrapper>
      </CloseIcon>
    </button>
    Dialog content
  </dialog>
</Dialog>
`;
